### DISCUSSION TITLE
Using custom validators

### DISCUSSION THREAD

**Discussion by BrandonMathis** [rmosolgo/graphql-ruby#3941] [Q&A] [answered]
In a rails app, I am writing a mutation that will use a custom validation to validate the format of [ISO 8601 duration strings](https://en.wikipedia.org/wiki/ISO_8601#Durations)

Code looks something like this

**app/graphql/app_schema.rb**
```ruby
class AppSchema < GraphQL::Schema
  mutation(Types::MutationType)
  query(Types::QueryType)
end
```

**app/graphql/types/mutation_type.rb**
```ruby
module Types
  class MutationType < Types::BaseObject
    field :create_resource, mutation: Mutations::CreateResource do
    end
  end
end
```

**app/graphql/mutations/create_resource.rb**
```ruby
module Mutations
  class CreateResource < BaseMutation
    argument :duration, String, required: false, validates: { iso_8601_duration: {} }
  end
end
```

My custom validation code looks something like this

```ruby
class ISO8601DurationValidator < GraphQL::Schema::Validator
  def validate(_object, _context, value)
    ActiveSupport::Duration.parse(value)
  rescue ActiveSupport::Duration::ISO8601Parser::ParsingError => e
    e.message
  end
end

GraphQL::Schema::Validator.install(:iso_8601_duration, ISO8601DurationValidator)
```

My question is, where in the file structure of a typical rails application should this custom validator code go? I originally put the class above my `AppSchema` and called `GraphQL::Schema::Validator.install` inside the `AppSchema` class but that raised an error ONLY when boot in production mode. This error does not appear in development mode

```
.../graphql-1.12.16/lib/graphql/schema/validator.rb:70:in `block in from_config': unknown validation: :iso_8601_duration (ArgumentError)
```

To avoid this error when i boot rails in production mode, i have to move the validator into my `initializers` which I would prefer not to do.

**Comment by rmosolgo** [comment]
> where 

It can go anywhere, as long as it follows Rails project structure conventions (https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#project-structure). For example: 

- `Iso8601DurationValidator` -> `app/graphql/iso_8601_duration_validator.rb`
- `Validators::Iso8601DurationValidator` -> `app/graphql/validators/iso_8601_duration_validator.rb`
- `Schema::Validators::Iso8601DurationValidator` -> `app/graphql/schema/validators/iso_8601_duration_validator.rb`

There's no convention with GraphQL-Ruby, so any of those should work fine. If you try one of those and find that it still doesn't work, please paste the full error message and stack trace here and we can look into it further!

**Comment by BrandonMathis** [comment]
Where specifically would you put this line of code?

```
GraphQL::Schema::Validator.install(:iso_8601_duration, ISO8601DurationValidator)
```

It makes sense to be to put that inside the class that defines the schema but doing this results an the following error

```
unknown validation: :iso_8601_duration
```

[Take a look @ my example rails project here](https://github.com/BrandonMathis/gql-validators) 😇

**Comment by rmosolgo** [comment, marked as answer]
I guess what counts is, that validator has to be registered _before_ any definitions actually use it. 

A couple options: 

- Put it in an initializer, so you can be sure that's run before any application code is loaded 
- Put it in `iso_8601_duration_validator.rb`, and add `require "graphql/(your path)/iso_8601_duration_validator.rb"` to any files that use that validator. That would make sure that validator was loaded and installed before those types are configured. 
- Probably, you could put that line before you schema class definition (eg `app/graphql/my_app_schema.rb`, before the class definition). That will work, _assuming_ that none of the types that use the validator are referenced outside the schema. (I have heard of applications that reference graphql types in general application logic -- not normal, not pretty, but it happens! If you don't have any cases of that, then adding that line to the top of the schema file should work.) 

If none of those work, I bet we can find something else.

**Comment by BrandonMathis** [comment]
Understood. For now, I think and initializer makes the most sense to me. Thankyou for the assistance.
